[OPTIMIZER][BACKEND] significantly cleaner handling of mixed-precision kernels

ptillet opened the pull request:
we currently have a very janky approach to optimizing mixed-precision matmul workloads, where some layout combinations (e.g., NT matmul) were explicitly pattern-matched to take a more optimized codepath. Attempt at unifying all the codepaths to codegen cp.async failed, due to bugs in SharedToDotOperandMMAv2.cpp.

This PR fixes said bugs, add some assertions for SharedToDotOperandMMAv2 modes that aren't well supported, and greatly simplify our handling of element-wise operations between load and conversions to DotOperand.

ptillet commented:
:warning: **This PR does not produce bitwise identical kernels as the branch it's merged against.** Please check artifacts for details. [Download the output file here](https://github.com/openai/triton/actions/runs/5605373713).

njsmith commented:
I've recently been trying to implement a kernel that does a int8 x bfloat16 matmul with triton, and isolated the issues I'm running into down to [a minimal example](https://gist.github.com/njsmith/174e21ccf1adc51413642fd5b13536c1). With current `main`, it silently gives incorrect results, but with this PR branch it actually works! So I guess I'm just writing this to cheer you on with this PR 🎉 

In case it's useful, I did notice that it still gives incorrect results if the A matrix has column major ordering. Probably not too important a case to support, but it'd be nice if it at least errored out instead of silently giving the wrong answer?

ptillet commented:
Yeah.. I think there were (and there may still be) pretty deep bugs in the way we emulate `ldmatrix` :x I've fixed some, and added some asserts, and also just generally cleaned things up.

Unfortunately this PR also has some performance regressions on some internal tests, so I still need to fix it before I can merge.

ptillet commented:
:warning: **This PR does not produce bitwise identical kernels as the branch it's merged against.** Please check artifacts for details. [Download the output file here](https://github.com/openai/triton/actions/runs/5638924365).
